text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+-   The new types do not implement the latest features implemented by the standard types.
+-   The new types do not implement all the traits implemented by the standard types, most especially 3rd-party traits.
+
+There are a few advantages to brand new types. For example, a nul-terminated in-line string does not have the 16 bytes
+overhead that a "naive" `String<InlineSingleStore<[u8; N]>>` would have. The Store API will not eliminate the potential
+need for such specialized collections, but it may eliminate the need for most alternatives in most situations.
+
+
+##  Allocator-like
+
+The API of `Store` is intentionally kept very close to that of `Allocator`.
+
+This similarility, and the similarity of the safety requirements, means that any developer accustomed to the current
+`Allocator` API can quickly dive in, and also means that bridging between `Store` and `Allocator` is as easy as
+possible.
+
+There are 3 extra pieces:
+
+-   The `Handle` associated type is used, rather than `NonNull<u8>`. This is the key to the flexibility of `Store`.
+-   The `dangling` method, reminiscent of `NonNull::dangling`, and to be used for the same purposes. It is part of
+    `Store` to simplify the overall API, as otherwise another trait would be required for `Handle`.
+-   The `resolve` method, which bridges from `Handle` to `NonNull<u8>`, since access to the allocated blocks of memory
+    require pointers to them.
+
+Otherwise, the bulk of the API is a straightforward translation of the `Allocator` API, substituting `Handle` anytime
+`NonNull<_>` appears.
+
+
+##  Guarantees, or absence thereof
+
+The `Store` API is minimalist, providing a minimum of guarantees.
+
+Beyond being untyped and unchecked, there are also a few oddities, compared to the `Allocator` API:
+
+-   By default, using `Store::allocate` or `Store::allocate_zeroed` invalidates all existing handles. This oddity
+    stems from the requirement of minimizing the state to be stored in collections using a single allocation at a time
+    such as `Box`, `Vec`, `VecDeque`, ...
+-   By default, calling any method -- including `resolve` -- invalidates all resolved pointers[^1]. This oddity stems
+    from the desire to leave the API flexible enough to allow caching stores, single-allocation stores, or copying
+    stores.
+-   By default, moving `Store` invalidates all resolved pointers. This oddity stems from the fact that when using an
+    in-line store the pointers point within the block of memory covered by the store, and thus after it moved, are left
+    pointing into the void.
+
+When the above should be guaranteed, extra marker traits can be implemented to provide compile-time checking that these
+properties hold, which in turn allows the final user to safely mix and match collection and store, relying on compiler
+errors to prevent erroneous couplings.
+
+[^1]: With the exception, in the case of a call to `resolve`, of any pointer derived from a copy of the handle argument.
```

Comment:
This flies _way_ over my head, I am afraid :/

Reply:
I think this actually only matters when the storage isn't modeled with shared mutability. It's the same restriction of only ever using pointers to the memory block instead of references.